iotilecore/iotile/core/hw/hwmanager.py

```diff
@@ -785,7 +788,47 @@ def get_proxy(self, short_name):
         if short_name not in self._name_map:
             return None
 
-        return self._name_map[short_name][0]
+        proxy_match = self.find_correct_proxy_version(self._name_map[short_name], version)
+        print("proxy_match: ", proxy_match)
+        return proxy_match if proxy_match is not None else self._name_map[short_name][0] 
+
+    def find_correct_proxy_version(self, proxies, version):
+        """Retrieves the ModuleVersion of each proxy and match it with the tile version"""
+
+        proxy_details = {}
+        tile_version = SemanticVersion(version[0], version[1], version[2])
+        latest_version = SemanticVersion(0, 0, 0)
+
+        for proxy in proxies:
+            proxy_details[proxy] = {}
+            try:
+                # If proxy has ModuleVersion(), get the SemanticVersionRange
+                module_version = proxy.ModuleVersion()
+                proxy_details[proxy]['ModuleVersion'] = module_version
+            except AttributeError:
+                # If proxy does not have ModuleVersion(), use the name of the support package
+                stripped_proxy_version = str(proxy).split('.')[0].split('_')[-1]
+                module_version = SemanticVersionRange.FromString('^' + stripped_proxy_version + '.0.0')
+                proxy_details[proxy]['ModuleVersion'] = module_version
+
+            # Keep track of each proxy's minimum version required
+            proxy_details[proxy]['MinVersion'] = module_version._disjuncts[0][0][0]
+
+            if module_version.check(tile_version):
+                # Mark the proxy as compatible, and check if it is the latest version of the proxy
+                proxy_details[proxy]['Compatible'] = True
+                min_version = proxy_details[proxy]['MinVersion']
+                latest_version = min_version if latest_version.__lt__(min_version) else latest_version
+            else:
+                proxy_details[proxy]['Compatible'] = False
+
+        # Return the latest/compatible proxy
+        for proxy, details in proxy_details.items():
+            if details['Compatible'] and details['MinVersion'] == latest_version:
+                return proxy
```

Comment:
I think you will need to iterate through the whole list and use a clamp (like recording the biggest number you've seen so far in a list) to select the "latest", as it is this returns on the first match instead of the latest match.

Reply:
I'm not sure that the "latest match" is necessarily the correct thing here, though. We may need to define what "best match" is -- essentially, how do we decide what to use when multiple compatible proxies are available? Should we choose the first one, or the one that's got the biggest version?

Reply:
Regarding your first comment... `latest_version` is checked in the first `for` loop, where it is updated as each proxy is visited. Not only that, but it marks itself as the latest version only when that proxy is compatible. That is why in the second `for` loop, we return right away when we see it because we've already iterated through the entire list of proxies.

Reply:
Regarding your second comment... In this case, "best match" is the "latest version" AND "compatible".